### PR TITLE
#615 miovision alert api

### PR DESCRIPTION
**Pull request by gabrielwol**
## What this pull request accomplishes:
- A small new DAG to pull Miovision Alerts into the database. Will be useful for communicating issues to Miovision - we can be aware of any current alerts. 
- Backfilled for this year. 
- Updated the readmes.

## Issue(s) this solves:
- #615 

## What, in particular, needs to reviewed:
- The insert_data sql. Updating existing records during the insert to extend end_times. 

## What needs to be done by a sysadmin after this PR is merged
- Checkout these files in data_scripts.

### CODE REVIEW

File: volumes/miovision/sql/table/create-table-alerts.sql [hunk 1]
```diff
@@ -0,0 +1,47 @@
+-- Table: miovision_api.alerts
+
+-- DROP TABLE IF EXISTS miovision_api.alerts;
+
+CREATE TABLE IF NOT EXISTS miovision_api.alerts
+(
+    intersection_id text COLLATE pg_catalog."default" NOT NULL,
```
[the remark this answers is not in the document]
Comment (gabrielwol) [reply volumes/miovision/sql/table/create-table-alerts.sql:7]: The reason I decided against adding the more usable `intersection_uid` to this table is that during my testing there were alerts for new intersections we don't have in our database yet + for this API endpoint there's no intersection param so we have to pull all the alerts. So if I add `intersection_uid` column then some will be null :(

File: volumes/miovision/sql/table/create-table-alerts.sql [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+GRANT SELECT ON TABLE miovision_api.alerts TO bdit_humans;
+
+GRANT ALL ON TABLE miovision_api.alerts TO miovision_admins;
+
+GRANT INSERT, SELECT, DELETE, UPDATE ON TABLE miovision_api.alerts TO miovision_api_bot;
+
+COMMENT ON TABLE miovision_api.alerts
+IS 'This table contains Miovision alerts to 5 minute accuracy,
+with maximum interval of 1 day. Pulled by a daily Airflow DAG `miovision_alerts`. 
+Note: a more detailed description is available on Miovision One.';
```
[the remark this answers is not in the document]
Comment (gabrielwol) [reply volumes/miovision/sql/table/create-table-alerts.sql:31]: oh good call!

File: dags/miovision_alerts.py [hunk 3]
```diff
@@ -0,0 +1,69 @@
+r"""### Daily Miovision Alert Pull DAG
```
[the remark this answers is not in the document]
Comment (radumas) [reply dags/miovision_alerts.py:1]: would we end up at a point where aggregation would depend on alert data?
I guess that would be more on the side of the views?

[the remark this answers is not in the document]
Comment (gabrielwol) [reply dags/miovision_alerts.py:1]: Requires some more investigation, but I think mostly alert == zero data.

File: volumes/miovision/sql/table/create-table-alerts.sql (hunk 1 above)
[the remark this answers is not in the document]
Comment (chmnata) [reply volumes/miovision/sql/table/create-table-alerts.sql:7]: HMMMM I see

[the remark this answers is not in the document]
Comment (chmnata) [reply volumes/miovision/sql/table/create-table-alerts.sql:7]: I think having null is fine in that case 🤔 , we can add a step in the adding intersection doc to do an update on this table after inserting new intersection? Having intersection_uid would be very helpful

File: volumes/miovision/sql/README.md [hunk 4]
```diff
@@ -417,6 +418,17 @@ centreline_id| numeric | Corresponds to `geo_id` in `gis.centreline`|14016757|
...
+
+**Field Name**|**Data Type**|**Description**|**Example**|
+:-----|:-----|:-----|:-----|
+intersection_id | text | The intersection id, corresponding to intersections.intersection_id column | c04704a0-e1e2-4101-9c29-6823d0f41c52 |
+alert | text | Short text description of the alert. Longer forms are available in the Miovision One UI | PERIPHERAL_UNAVAILABLE |
+start_time | timestamp | First 5 minute interval at which the alert appeared. **Subtract 5 minutes to get earliest possible start time.** | 2024-01-12 10:20:00 | 
+end_time | timestamp | Final 5 minute interval at which the alert appeared. **Add 5 minutes to get latest possible end time.** Note this could be extended the following day. | 2024-01-21 15:35:00 | 
+
```
[the remark this answers is not in the document]
Comment (chmnata) [reply volumes/miovision/sql/README.md:431]: nvm, the open_issue table is a better candidate for this

File: dags/miovision_alerts.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (gabrielwol) [reply dags/miovision_alerts.py:1]: I was wrong! There are volumes during many of these alerts! We may need to use these similarly to anomalous_ranges. Will open a new issue. 

```sql
SELECT a.intersection_uid, alert, start_time, end_time, SUM(volume)
FROM miovision_api.alerts AS a
LEFT JOIN miovision_api.volumes_2024 AS v
    ON a.intersection_uid = v.intersection_uid
    AND v.datetime_bin >= a.start_time
    AND v.datetime_bin < a.end_time
WHERE a.intersection_uid IS NOT NULL
GROUP BY a.intersection_uid, alert, start_time, end_time
```

| "intersection_uid" | "alert"                  | "start_time"          | "end_time"            | "sum"   |
|--------------------|--------------------------|-----------------------|-----------------------|---------|
| 7                  | "PERIPHERAL_UNAVAILABLE" | "2024-02-14 01:05:00" | "2024-02-14 06:00:00" | 2247    |
| 24                 | "PERIPHERAL_UNAVAILABLE" | "2024-02-12 16:55:00" | "2024-02-12 17:20:00" | 1370    |
| 67                 | "TELEMETRY_UNAVAILABLE"  | "2024-01-01 00:00:00" | "2024-04-18 00:00:00" | 5936558 |
| 24                 | "PERIPHERAL_UNAVAILABLE" | "2024-02-11 02:00:00" | "2024-02-11 02:55:00" | 725     |
| 18                 | "POWER_OFF"              | "2024-04-10 02:10:00" | "2024-04-10 02:10:00" |         |
| 59                 | "PERIPHERAL_UNAVAILABLE" | "2024-03-24 01:45:00" | "2024-03-24 03:25:00" | 341     |
| 55                 | "POWER_OFF"              | "2024-03-19 14:45:00" | "2024-03-19 14:50:00" |         |
| 24                 | "PERIPHERAL_UNAVAILABLE" | "2024-02-10 01:40:00" | "2024-02-10 02:35:00" | 823     |
| 28                 | "POWER_OFF"              | "2024-03-13 10:00:00" | "2024-03-13 10:50:00" |         |
| 7                  | "IP_CONFLICT"            | "2024-02-06 13:30:00" | "2024-02-06 13:40:00" | 497     |

File: volumes/miovision/sql/README.md [hunk 5]
```diff
@@ -417,6 +418,17 @@ centreline_id| numeric | Corresponds to `geo_id` in `gis.centreline`|14016757|
...
+**Field Name**|**Data Type**|**Description**|**Example**|
+:-----|:-----|:-----|:-----|
+intersection_id | text | The intersection id, corresponding to intersections.intersection_id column | c04704a0-e1e2-4101-9c29-6823d0f41c52 |
+alert | text | Short text description of the alert. Longer forms are available in the Miovision One UI | PERIPHERAL_UNAVAILABLE |
+start_time | timestamp | First 5 minute interval at which the alert appeared. **Subtract 5 minutes to get earliest possible start time.** | 2024-01-12 10:20:00 | 
+end_time | timestamp | Final 5 minute interval at which the alert appeared. **Add 5 minutes to get latest possible end time.** Note this could be extended the following day. | 2024-01-21 15:35:00 |
```
[the remark this answers is not in the document]
Comment (gabrielwol) [reply volumes/miovision/sql/README.md:431]: Example added!

**Review by gabrielwol — COMMENTED** [review]
Review changes made!

File: volumes/miovision/sql/table/create-table-alerts.sql (hunk 2 above)
[the remark this answers is not in the document]
Comment (chmnata) [reply volumes/miovision/sql/table/create-table-alerts.sql:36]: peffffecccgt

File: volumes/miovision/api/pull_alert.py [hunk 6]
```diff
@@ -0,0 +1,180 @@
...
+
+    logger.info('Pulling Miovision alerts from %s to %s.', start_date, end_date)
+
+    #pull alerts at each timestamps and append to list
+    dfs = []
+    for dt in dt_list: 
+        miovpull = MiovAlertPuller(dt, key)
+        test = miovpull.get_response()
+        response = miovpull.process_response(test)
+        df = pd.DataFrame(response).drop_duplicates()
```
[the remark this answers is not in the document]
Comment (chmnata) [reply volumes/miovision/api/pull_alert.py:156]: got it!

**Review by chmnata — CHANGES_REQUESTED** [review]
One last change before merging

**Review by chmnata — APPROVED** [review]

**Pull request merged**